jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ScanningAppProvider.java

```diff
@@ -120,91 +85,66 @@ public void setUseRealPaths(boolean useRealPaths)
         _useRealPaths = useRealPaths;
     }
 
+    protected List<Unit> getUnitsForEnvironment(String envName)
+    {
+        return _units.getUnits().stream()
+            .filter((unit) -> envName.equals(unit.getEnvironmentName()))
+            .toList();
+    }
+
     protected void setFilenameFilter(FilenameFilter filter)
     {
         if (isRunning())
             throw new IllegalStateException();
         _filenameFilter = filter;
     }
 
+    public Comparator<Unit> getUnitComparator()
+    {
+        return _unitComparator;
+    }
+
+    public void setUnitComparator(Comparator<Unit> comparator)
+    {
+        this._unitComparator = comparator;
+    }
+
     /**
      * @return The index of currently deployed applications.
      */
-    protected Map<Path, App> getDeployedApps()
+    protected Collection<App> getDeployedApps()
     {
-        return _appMap;
+        return _units.getUnits()
+            .stream()
+            .map(Unit::getApp)
+            .collect(Collectors.toUnmodifiableSet());
     }
 
     /**
      * Called by the Scanner.DiscreteListener to create a new App object.
      * Isolated in a method so that it is possible to override the default App
      * object for specialized implementations of the AppProvider.
      *
-     * @param path The file that is the context.xml. It is resolved by
-     * {@link org.eclipse.jetty.util.resource.ResourceFactory#newResource(String)}
-     * @return The App object for this particular context definition file.
+     * @param path The file that the main point of deployment (eg: a context XML, a WAR file, a directory, etc)
+     * @return The App object for this particular context.
      */
     protected App createApp(Path path)
```

Comment:
If the `Unit` produced the `App` then that would better control that the `path` must be one of those known about by the `Unit`. In fact, you should probably be able to navigate from an `App` to the parent `Unit` also.

Reply:
Not all `Unit` instances have `App`. (re: Environment Config files are grouped into a `Unit` by its environment name, and trigger synthetic Changes to all registered `App` that are managed by `ContextProvider` and being tracked by our `Unit`).  

`App` instances from other (custom) `AppProviders` do not have a `Unit` and are not seen by our `ContextProvider`.

Only our `ContextProvider` needs to know the `App` for the `Unit`.
Keep in mind that `Unit` is just ours, not a general purpose API, like `App` is.

Reply:
I feel it is wrong that both App and Unit persist.    App should be the long lived result of an stream of `Unit` events.   Having them both persist is confusing